### PR TITLE
Upgrade Operator to 3.0.1

### PR DESCRIPTION
**Pull request by dvaldivia**
Fixes #199

### CODE REVIEW

File: restapi/admin_tenants.go [hunk 1]
```diff
@@ -134,7 +134,7 @@ func getDeleteTenantResponse(session *models.Principal, params admin_api.DeleteT
 	return deleteTenantAction(context.Background(), opClient, params.Namespace, params.Tenant)
 }
 
-func getMinioInstanceScheme(mi *operator.MinIOInstance) string {
+func getMinioInstanceScheme(mi *operator.Tenant) string {
```
Comment (cesnietor) [inline restapi/admin_tenants.go:137]: please update the name of the function

File: restapi/admin_tenants.go [hunk 2]
```diff
@@ -169,46 +169,48 @@ func getTenantAdminClient(ctx context.Context, client K8sClient, namespace, tena
 	return mAdmin, nil
 }
 
-func getMinioInstance(ctx context.Context, operatorClient OperatorClient, namespace, tenantName string) (*operator.MinIOInstance, error) {
-	minInst, err := operatorClient.MinIOInstanceGet(ctx, namespace, tenantName, metav1.GetOptions{})
+func getMinioInstance(ctx context.Context, operatorClient OperatorClient, namespace, tenantName string) (*operator.Tenant, error) {
```
Comment (cesnietor) [inline restapi/admin_tenants.go:172]: please update the name of the function

File: restapi/admin_tenants.go [hunk 3]
```diff
@@ -385,54 +386,28 @@ func getTenantCreatedResponse(session *models.Principal, params admin_api.Create
...
-				corev1.ResourceMemory:  memorySize,
-			},
-		},
-	}
-
-	if params.Body.VolumeConfiguration.StorageClass != "" {
-		volTemp.StorageClassName = &params.Body.VolumeConfiguration.StorageClass
-	}
-
 	//Construct a MinIO Instance with everything we are getting from parameters
```
Comment (cesnietor) [inline restapi/admin_tenants.go:399]: change comment

File: restapi/admin_tenants_test.go [hunk 4]
```diff
@@ -371,7 +370,7 @@ func Test_deleteTenantAction(t *testing.T) {
 		},
 	}
 	for _, tt := range tests {
-		opClientMinioInstanceDeleteMock = tt.args.mockMinioInstanceDelete
+		opClientTenantDeleteMock = tt.args.mockMinioInstanceDelete
```
Comment (cesnietor) [inline restapi/admin_tenants_test.go:373]: please update the args.Mock

File: restapi/admin_tenants_test.go [hunk 5]
```diff
@@ -461,8 +460,8 @@ func Test_TenantAddZone(t *testing.T) {
 		},
 	}
 	for _, tt := range tests {
-		opClientMinioInstanceGetMock = tt.args.mockMinioInstanceGet
-		opClientMinioInstancePatchMock = tt.args.mockMinioInstancePatch
+		opClientTenantGetMock = tt.args.mockMinioInstanceGet
```
Comment (cesnietor) [inline restapi/admin_tenants_test.go:463]: please update mockName

File: restapi/admin_tenants_test.go [hunk 6]
```diff
@@ -621,8 +620,8 @@ func Test_UpdateTenantAction(t *testing.T) {
...
```
Comment (cesnietor) [inline restapi/admin_tenants_test.go:623]: update mockname

**Comment by cesnietor** [comment]
please also update the kustomize files if needed.

File: restapi/admin_tenants.go [hunk 7]
```diff
@@ -460,10 +435,11 @@ func getTenantCreatedResponse(session *models.Principal, params admin_api.Create
 			return nil, err
 		}
 
-		minInst.Spec.MCS = &operator.MCSConfig{
-			Replicas:  2,
-			Image:     "minio/mcs:v0.1.1",
-			MCSSecret: &corev1.LocalObjectReference{Name: mcsSecretName},
+		const consoleVersion = "minio/mcs:v0.1.1"
```
[the remark this answers is not in the document]
Comment (dvaldivia) [reply restapi/admin_tenants.go:438]: I did it, then extracted to constant, I'll update it

File: k8s/operator-console/base/minio-operator.yaml [hunk 8]
```diff
@@ -192,12 +275,8 @@ spec:
...
-      serviceAccountName: minio-operator
       containers:
-        - name: minio-operator
-          image: minio/k8s-operator:2.0.9
+        - image: minio/k8s-operator:3.0.0
           imagePullPolicy: IfNotPresent
-          # To specify cluster domain, un comment the following:
-          # env:
-          #   - name: CLUSTER_DOMAIN
-          #     value: mycluster.mydomain
```
Comment (Alevsk) [inline k8s/operator-console/base/minio-operator.yaml:203]: is this env not supported anymore?

Comment (dvaldivia) [reply k8s/operator-console/base/minio-operator.yaml:203]: it is, but if wanted they should add it themselves, It's still on the kustomize files

**Comment by dvaldivia** [comment]
@bexsoft fixed

**Comment by cesnietor** [comment]
please make it a single commit

**Comment by dvaldivia** [comment]
@cesnietor done

File: swagger.yml [hunk 9]
```diff
@@ -1762,7 +1749,6 @@ definitions:
     type: object
     required:
       - name
-      - volume_configuration
```
Comment (cesnietor) [inline swagger.yml:1765]: Zone should be `required` else:
```
{
    "code": 500,
    "message": "Tenant.minio.min.io \"minio-tenant-1\" is invalid: spec.zones: Invalid value: \"null\": spec.zones in body must be of type array: \"null\""
}
```
Sending:
```
{
    "name": "minio-tenant-1",
    "namespace": "default"
}
```
``

**Comment by dvaldivia** [comment]
@cesnietor fixed

**Comment by bexsoft** [comment]
Socket hangup trying to gather tenant information. (http://localhost:9090/api/v1/namespaces/default/tenants/minio-tenant-4)

```
goroutine 144 [running]:                                                                                                           │
│ net/http.(*conn).serve.func1(0xc0004fdea0)                                                                                         │
│     /usr/local/go/src/net/http/server.go:1767 +0x139                                                                               │
│ panic(0x1ae5fc0, 0x3180a00)                                                                                                        │
│     /usr/local/go/src/runtime/panic.go:679 +0x1b2                                                                                  │
│ github.com/minio/mcs/restapi.getTenantInfo(0xc000c94240, 0xc0007c1fe0, 0x213c880)                                                  │
│     /go/src/github.com/minio/mcs/restapi/admin_tenants.go:198 +0x3a3                                                               │
│ github.com/minio/mcs/restapi.getTenantInfoResponse(0xc0005c0870, 0xc000a8a800, 0xc000246217, 0x7, 0xc000246227, 0xe, 0x0, 0x0, 0x0 │
│     /go/src/github.com/minio/mcs/restapi/admin_tenants.go:259 +0x653                                                               │
│ github.com/minio/mcs/restapi.registerTenantHandlers.func4(0xc000a8a800, 0xc000246217, 0x7, 0xc000246227, 0xe, 0xc0005c0870, 0x1ca1 │
│     /go/src/github.com/minio/mcs/restapi/admin_tenants.go:77 +0x59                                                                 │
│ github.com/minio/mcs/restapi/operations/admin_api.TenantInfoHandlerFunc.Handle(0x1e6c110, 0xc000a8a800, 0xc000246217, 0x7, 0xc0002 │
│     /go/src/github.com/minio/mcs/restapi/operations/admin_api/tenant_info.go:38 +0x51                                              │
│ github.com/minio/mcs/restapi/operations/admin_api.(*TenantInfo).ServeHTTP(0xc000481960, 0x20ed5c0, 0xc0002762a0, 0xc000a8a800)     │
│     /go/src/github.com/minio/mcs/restapi/operations/admin_api/tenant_info.go:86 +0x342                                             │
│ github.com/go-openapi/runtime/middleware.NewOperationExecutor.func1(0x20ed5c0, 0xc0002762a0, 0xc000a8a700)                         │
│     /go/pkg/mod/github.com/go-openapi/runtime@v0.19.19/middleware/operation.go:28 +0x75                                            │
│ net/http.HandlerFunc.ServeHTTP(0xc0006c17d0, 0x20ed5c0, 0xc0002762a0, 0xc000a8a700)                                                │
│     /usr/local/go/src/net/http/server.go:2007 +0x44                                                                                │
│ github.com/go-openapi/runtime/middleware.NewRouter.func1(0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                    │
│     /go/pkg/mod/github.com/go-openapi/runtime@v0.19.19/middleware/router.go:77 +0x356                                              │
│ net/http.HandlerFunc.ServeHTTP(0xc000d157e0, 0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                                │
│     /usr/local/go/src/net/http/server.go:2007 +0x44                                                                                │
│ github.com/go-openapi/runtime/middleware.Redoc.func1(0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                        │
│     /go/pkg/mod/github.com/go-openapi/runtime@v0.19.19/middleware/redoc.go:72 +0x286                                               │
│ net/http.HandlerFunc.ServeHTTP(0xc000cddb00, 0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                                │
│     /usr/local/go/src/net/http/server.go:2007 +0x44                                                                                │
│ github.com/go-openapi/runtime/middleware.Spec.func1(0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                         │
│     /go/pkg/mod/github.com/go-openapi/runtime@v0.19.19/middleware/spec.go:46 +0x188                                                │
│ net/http.HandlerFunc.ServeHTTP(0xc000cddb40, 0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                                │
│     /usr/local/go/src/net/http/server.go:2007 +0x44                                                                                │
│ github.com/minio/mcs/restapi.FileServerMiddleware.func1(0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                     │
│     /go/src/github.com/minio/mcs/restapi/configure_mcs.go:184 +0x1e2                                                               │
│ net/http.HandlerFunc.ServeHTTP(0xc000d15c80, 0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                                │
│     /usr/local/go/src/net/http/server.go:2007 +0x44                                                                                │
│ github.com/unrolled/secure.(*Secure).Handler.func1(0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                          │
│     /go/pkg/mod/github.com/unrolled/secure@v1.0.7/secure.go:177 +0xdf                                                              │
│ net/http.HandlerFunc.ServeHTTP(0xc000d15ca0, 0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                                │
│     /usr/local/go/src/net/http/server.go:2007 +0x44                                                                                │
│ net/http.serverHandler.ServeHTTP(0xc000ce69a0, 0x20ed5c0, 0xc0002762a0, 0xc000a8a500)                                              │
│     /usr/local/go/src/net/http/server.go:2802 +0xa4                                                                                │
│ net/http.(*conn).serve(0xc0004fdea0, 0x20f4f00, 0xc0005d6140) ...
```

Screen Shot 2020-07-24 at 12 51 40 PM

**Comment by cesnietor** [comment]
Since A tenant can be created without serviceName, it now fails getting Tenant's info:
on GET:
```
{
    "code": 500,
    "message": "resource name may not be empty"
}
```
caused by :
```
error getting tenant's admin client: resource name may not be empty        
```

**Comment by cesnietor** [comment]
Tenants are never created:
```
0-07-24T17:40:44.4224816Z I0724 17:40:44.411016       1 main-controller.go:320] Successfully synced 'default/minio-tenant-2'                                                                                                   │
│ 2020-07-24T17:40:44.4511057Z E0724 17:40:44.450764       1 main-controller.go:325] error syncing 'default/minio-tenant-2': Endpoint: :9000 does not follow ip address or domain name standards.      
 2020-07-24T17:37:29.0664031Z I0724 17:37:29.065859       1 main-controller.go:320] Successfully synced 'default/minio-tenant-1'                                                                                                   │
│ 2020-07-24T17:37:29.0881936Z E0724 17:37:29.087572       1 main-controller.go:325] error syncing 'default/minio-tenant-1': Endpoint: :9000 does not follow ip address or domain name standards.   
```

**Comment by cesnietor** [comment]
Since this changes will affect UI it needs to also be addressed in this PR.

**Comment by cesnietor** [comment]
with the following request the minio pods are never starting:
```
{
    "name": "minio-tenant-2",
    "namespace": "default",
    "zones": [
        {
            "servers": 2,
            "volumes_per_server": 1,
            "volume_configuration": {
                "storage_class_name": "standard",
                "size": 1073741824
            }
        }
    ]
}
```
```
ERROR Invalid command line arguments: Incorrect number of endpoints provided [http://minio-tenant-2-zone-0-{0...1}.minio-tenant-2-hl.default.svc.cluster.local/export]
      > Please provide correct combination of local/remote paths
      HINT:
        For more information, please refer to https://docs.min.io/docs/minio-erasure-code-quickstart-guide
```

File: k8s/operator-console/base/minio-operator.yaml [hunk 10]
```diff
@@ -276,7 +276,7 @@ spec:
         name: minio-operator
     spec:
       containers:
-        - image: minio/k8s-operator:3.0.0
+        - image: minio/operator:v3.0.1
```
Comment (harshavardhana) [inline k8s/operator-console/base/minio-operator.yaml:279]: container name is still k8s-operator @dvaldivia

Comment (dvaldivia) [reply k8s/operator-console/base/minio-operator.yaml]: it was reverted

**Comment by cesnietor** [comment]
If I create a Tenant, then I delete it (with its secrets) and then Create it again with the same request: the tenant doesn't start:
and the pods log:
```
│ 2020-07-25T21:05:18.7991578Z ERROR Unable to initialize server switching into safe-mode: Unable to initialize sub-systems: Unable to initialize config system: Invalid credentials                                                │
│                                                                                                                                                                                              
```
Request used:
```
{
    "name": "minio-tenant-3",
    "namespace": "default",
    "zones": [
        {
            "servers": 1,
            "volumes_per_server": 4,
            "volume_configuration": {
                "storage_class_name": "standard",
                "size": 1073741824
            }
        }
    ]
}
```
steps:
1-Create
2-DElete tenant
3- Delete secrets
4 - Repeat 1

**Comment by harshavardhana** [comment]
> If I create a Tenant, then I delete it (with its secrets) and then Create it again with the same request: the tenant doesn't start:
> and the pods log:
> 
> ```
> │ 2020-07-25T21:05:18.7991578Z ERROR Unable to initialize server switching into safe-mode: Unable to initialize sub-systems: Unable to initialize config system: Invalid credentials                                                │
> │                                                                                                                                                                                              
> ```
> 
> Request used:
> 
> ```
> {
>     "name": "minio-tenant-3",
>     "namespace": "default",
>     "zones": [
>         {
>             "servers": 1,
>             "volumes_per_server": 4,
>             "volume_configuration": {
>                 "storage_class_name": "standard",
>                 "size": 1073741824
>             }
>         }
>     ]
> }
> ```
> 
> steps:
> 1-Create
> 2-DElete tenant
> 3- Delete secrets
> 4 - Repeat 1

Secrets cannot be deleted, secrets are used for encrypting the config and IAM assets in MinIO so it is failing correctly.

Looks like an operator issue.

**Comment by cesnietor** [comment]
> Secrets cannot be deleted

I meant the k8s secrets that mcs create

**Comment by harshavardhana** [comment]
> > Secrets cannot be deleted
> 
> I meant the k8s secrets that mcs create

If the same disks are used, same secrets should be used - I am explaining what is required from MinIO.

**Comment by cesnietor** [comment]
> Looks like an operator issue.

I will approve this since we can update mcs with this minor bug and update once operator has the fix.

**Comment by harshavardhana** [comment]
So it a tenant is deleted and somehow new secrets get added it is not allowed to start.

https://github.com/minio/minio/tree/master/docs/config#rotating-encryption-with-new-credentials

**Comment by cesnietor** [comment]
> So it a tenant is deleted and somehow new secrets get added it is not allowed to start.
> 
> https://github.com/minio/minio/tree/master/docs/config#rotating-encryption-with-new-credentials

Oh I see.

**Pull request merged**